src/aes.rs

```diff
@@ -0,0 +1,267 @@
+// Copyright (c) 2022, Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::marker::PhantomData;
+
+use crate::traits::{AuthenticatedCipher, Cipher, EncryptionKey, ToFromBytes};
+
+use aes::cipher::{BlockDecryptMut, BlockEncryptMut};
+use aes_gcm::{AeadInPlace, KeyInit};
+use ctr::cipher::StreamCipher;
+use digest::{crypto_common::KeyIvInit, generic_array::ArrayLength, typenum::U16};
+use serde::{Deserialize, Serialize};
+use serde_with::serde_as;
+
+///
+/// An encryption key for a symmetric cipher consisting of `N` bytes.
+///
+#[serde_as]
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct Key<const N: usize> {
+    // This is needed to derive Serialize and Deserialize since serde does not support derive for arrays
+    // longer than 32 or of generic size. We do not really need it since N currently is either 16 or 32
+    // but it allows generifying the code.
+    #[serde_as(as = "[_; N]")]
+    pub bytes: [u8; N],
+}
+
+impl<const N: usize> AsRef<[u8]> for Key<N> {
+    fn as_ref(&self) -> &[u8] {
+        self.bytes.as_ref()
+    }
+}
+
+impl<const N: usize> ToFromBytes for Key<N> {
+    fn from_bytes(bytes: &[u8]) -> Result<Self, signature::Error> {
+        Ok(Key {
+            bytes: bytes.try_into().expect("wrong length"),
+        })
+    }
+
+    fn as_bytes(&self) -> &[u8] {
+        &self.bytes
+    }
+}
+
+impl<const N: usize> EncryptionKey for Key<N> {
+    fn generate<R: rand::CryptoRng + rand::RngCore>(rng: &mut R) -> Self {
+        let mut bytes = [0u8; N];
+        rng.fill_bytes(&mut bytes);
+        Key { bytes }
+    }
+}
+
+///
+/// Aes in CTR mode
+///
+pub struct AesCtr<const KEY_SIZE: usize, Aes> {
+    key: Key<KEY_SIZE>,
+    algorithm: PhantomData<Aes>,
+}
+
+impl<const KEY_SIZE: usize, Aes> AesCtr<KEY_SIZE, Aes> {
+    pub fn new(key: Key<KEY_SIZE>) -> Self {
+        Self {
+            key,
+            algorithm: PhantomData,
+        }
+    }
+}
+
+impl<const KEY_SIZE: usize, Aes> Cipher for AesCtr<KEY_SIZE, Aes>
+where
+    Aes: aes::cipher::KeySizeUser
+        + aes::cipher::KeyInit
+        + aes::cipher::BlockCipher
+        + aes::cipher::BlockSizeUser<BlockSize = U16>
+        + aes::cipher::BlockEncrypt
+        + aes::cipher::BlockDecrypt,
+{
+    fn encrypt(
+        &self,
+        iv: &[u8],
```

Comment:
A few questions:
- do we already check iv size mismatch?
- we handle the AES `Key` as a special struct, what about the `IV`? maybe for consistency we should as well. Not strongly opinionated on that.
- For all of the encrypt/decrypt and their authenticated variants, let's have some tests on input [u8] sizes mismatch.